enterprise/internal/campaigns/rewirer/rewirer.go

```diff
@@ -181,12 +182,21 @@ func (r *ChangesetRewirer) closeChangeset(changeset *campaigns.Changeset) {
 				changeset.PreviousSpecID = changeset.CurrentSpecID
 			}
 			changeset.Closing = true
-			changeset.ResetQueued()
+			reset = true
 		}
 	}
 
 	// Disassociate the changeset with the campaign.
-	changeset.RemoveCampaignID(r.campaignID)
+	for i := range changeset.Campaigns {
+		if changeset.Campaigns[i].CampaignID == r.campaignID && !changeset.Campaigns[i].Detach {
+			changeset.Campaigns[i].Detach = true
+			reset = true
+			break
+		}
+	}
+	if reset {
+		changeset.ResetQueued()
+	}
```

Comment:
How about changing the signature of `RemoveCampaignID` to return a `bool`, representing whether the given ID was removed or not? 

Then you could do

```suggestion
	if ok := changeset.RemoveCampaignID(r.campaignID); ok {
		reset = true
	}
	if reset {
		changeset.ResetQueued()
	}
```

Reply:
Cool idea, but I actually want to mark is as "Detach", not remove it from the slice, so the use case is slightly different here.